core/src/main/scala/org/apache/spark/sql/catalyst/expressions/ParamLiteral.scala

```diff
@@ -146,23 +180,74 @@ case class ParamLiteral(l: Literal, pos: Int) extends LeafExpression {
     }
   }
 
-  override def nullable: Boolean = l.nullable
+}
 
-  override def eval(input: InternalRow): Any = l.eval()
+object ParamLiteral {
+  def apply(_value: Any, _dataType: DataType, pos: Int): ParamLiteral =
+    new ParamLiteral(_value, _dataType, pos)
 
-  override def dataType: DataType = l.dataType
+  def unapply(arg: ParamLiteral): Option[(Any, DataType, Int)] =
+    Some((arg.value, arg.dataType, arg.pos))
 }
 
-case class LiteralValue(var value: Any, var position: Int)
-  extends KryoSerializable {
+case class LiteralValue(var value: Any, var dataType: DataType, var position: Int)
+    (var converter: Any => Any = createToScalaConverter(dataType))
+    extends KryoSerializable {
 
   override def write(kryo: Kryo, output: Output): Unit = {
     kryo.writeClassAndObject(output, value)
+    kryo.writeClassAndObject(output, dataType.jsonValue)
     output.writeVarInt(position, true)
   }
 
   override def read(kryo: Kryo, input: Input): Unit = {
     value = kryo.readClassAndObject(input)
+    dataType = DataType.parseDataType(kryo.readClassAndObject(input).asInstanceOf[JValue])
     position = input.readVarInt(true)
+    converter = createToScalaConverter(dataType)
+  }
+}
+
+
+/**
+ * Wrap any ParamLiteral expression with this so that we can generate literal initialization code
+ * within the <code>.init()</code> method of the generated class.
+ * <br><br>
+ *
+ * We try to locate first foldable expression in a query tree such that all its child is foldable
+ * but parent isn't. That way we locate the exact point where an expression is safe to evalute
+ * once instead of evaluating every row.
+ * <br><br>
+ *
+ * Expressions like <code> select c from tab where
+ *  case col2 when 1 then col3 else 'y' end = 22 </code>
+ * like queries doesn't converts literal evaluation into init method.
+ *
+ * @param expr minimal expression tree that can be evaluated only once and turn into a constant.
+ */
+case class DynamicFoldableExpression(expr: Expression) extends Expression {
+  override def nullable: Boolean = expr.nullable
+
+  override def eval(input: InternalRow): Any = expr.eval(input)
+
+  override protected def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    val eval = expr.genCode(ctx)
+    val newVar = ctx.freshName("paramLiteralExpr")
+    val newVarIsNull = ctx.freshName("paramLiteralExprIsNull")
+    val comment = ctx.registerComment(expr.toString)
+    ctx.addMutableState(ctx.javaType(expr.dataType), newVar,
+      s"$comment\n${eval.code}\n$newVar = ${eval.value};")
+    ctx.addMutableState("boolean", newVarIsNull, s"$newVarIsNull = ${eval.isNull};")
```

Comment:
I think this code for both newVar and newVarIsNull should be put inside a separate function instead of cluttering the init method. We also have a problem of 64K method limit so that will be better since "eval.code" can potentially be many and long.